### PR TITLE
helper/schema: Allow self references with AllowOneOf, ExactlyOneOf, and RequiredWith in InternalValidate

### PR DESCRIPTION
**Pull request by bflad**
Reference: #407
Reference: #416

### CODE REVIEW

File: helper/schema/schema.go [hunk 1]
```diff
@@ -948,6 +954,11 @@ func checkKeysAgainstSchemaFlags(k string, keys []string, topSchemaMap schemaMap
 			return fmt.Errorf("%s cannot contain Computed(When) attribute (%s)", k, key)
 		}
 	}
+
+	if requireSelfReference && !foundSelfReference {
```
Comment (appilon) [inline helper/schema/schema.go:958]: `AtLeastOneOf` and `ExactlyOneOf` were implemented with an idempotent line that adds itself and removes duplicate, it allows the self reference to be optional. I would change this PR to have a flag for `allowSelfRef` and ditch the requirement for it, but rather disallow for `ConflictsWith` (and possibly `RequiredWith`)

Comment (appilon) [reply helper/schema/schema.go:958]: Unless we remove that line in `validateRequiredWith` I would actually ONLY have this error for `ConflictsWith`, just to be consistent, I will approve either final implementation.

**Comment by paultyng** [comment]
Per Alex's comment, may be worth referencing back to the RFC, there was some discussion about the requirement of the self reference, I think we landed on not requiring it for a reason, though I can't recall the reason 😂

**Comment by bflad** [comment]
I'll update this to be allow vs require and add `RequiresWhen` unit testing as well to ensure that new attribute is covered since its not currently (with allowed self-references for it). 👍

**Comment by bflad** [comment]
Implementation updated and submitted v1 backport: #423

**Pull request merged**